Wait countdown

Purplo-cf opened the pull request:
Adds support for new WaitCountdown prefab in 5-fret, drums, and vocals engines during long stretches with no notes

sonicfind commented on Assets/WaitCountdown.cs:
```diff
@@ -0,0 +1,78 @@
+                        gameObject.SetActive(true);
+                    }
+                }
+                else
+                {
+                    //hide countdown when there is one measure before the next set of notes
+                    gameObject.SetActive(false);
+                }
```
Early return here as well (we'll figure out how to make it fade out in the future)

Purplo-cf replied:
Nevermind, it was simple enough to copy it over. I kept the added fade animation in its own commit so it could be reverted if you'd rather not have it

sonicfind commented on Assets/WaitCountdown.cs:
```diff
@@ -0,0 +1,78 @@
+                    GameObject childToShow = null;
+                    GameObject childToHide = null;
+
+                    if (measuresLeft >= 2)
+                    {
+                        _countdownText.text = measuresLeft.ToString();
+
+                        childToHide = _getReady.gameObject;
+                        childToShow = _backgroundCircle.gameObject;
+                    }
+                    else if (measuresLeft == 1)
+                    {
+                        childToHide = _backgroundCircle.gameObject;
+                        childToShow = _getReady.gameObject;
+                    }
```
Firstly, much rather have the active state directly set appropriately within these scopes over holding the objects in temporary variables.

Secondly, the `measuresLeft == 1` conditional is redundant because we already checked for `measuresLeft > 0` before this point.

Purplo-cf replied:
Ready for another look

sonicfind commented on Assets/WaitCountdown.cs:
```diff
@@ -0,0 +1,78 @@
+                    }
+
+                    //update visibility of child objects
+                    if (childToShow?.activeSelf == false)
+                    {
+                        childToShow.SetActive(true);
+                    }
+
+                    if (childToHide?.activeSelf == true)
+                    {
+                        childToHide.SetActive(false);
+                    }
+
+                    //update visibility of parent object if hidden
+                    if (!gameObject.activeSelf)
+                    {
+                        gameObject.SetActive(true);
+                    }
```
I believe these checks of the `activeSelf` property are redundant. Unity should already run said test before making changes to the gameObject's state. Just running `SetActive` on it's own should suffice.

EliteAsian123 replied:
Just to add on, using `?` on Unity objects is generally not advised since they treat null differently. It's best to check if it's null explicitly using `==` (which is slow by the way, because Unity).

Purplo-cf replied:
Ready for another look

sonicfind reviewed (changes requested):
Just some minor syntactical changes. The functionality looks to be there (though I haven't actually tested it myself).

Purplo-cf commented on Assets/WaitCountdown.cs:
```diff
@@ -0,0 +1,78 @@
+            if (measuresLeft != _measuresLeft)
+            {
+                _measuresLeft = measuresLeft;
+
```
Ready for another look

sonicfind commented:
![image](https://github.com/YARC-Official/YARG/assets/52436993/37870a33-0093-4828-9480-e2ae70b84214)
I would make this smaller. Having it span the entire width of the highway is a bit much.
The height also doesn't match the height of the circle, so there's a bit of a disconnect.

sonicfind commented:
Also, I'll mention this here too: Don't *merge*, **rebase**. Because of all you're merging, your PR is mostly now just a bunch of merge commits. It makes it nearly impossible now to see what changes to your own code were made in the sea of the merged commits.

Purplo-cf commented:
I have adjusted the size of the "Get Ready!" box and added a hook for it into the localization table. Pinging @kaduwaengertner since he mentioned that he already created assets for this. Are those ready to implement?

Also, I just realized that Clone Hero only shows the the countdown and doesn't do the "Get Ready!" that RB4 does. Do you have a preference between the two styles?

If it helps, this was the original RB4 reference that I used for for sizing and positioning
![RB4_Countdown_ref](https://github.com/YARC-Official/YARG/assets/33099317/d247cef6-4e25-44d4-af91-6559b9410707)

kaduwaengertner commented:
Just added a reminder for myself to send you the assets tomorrow when Im at my pc.

TheNathannator commented:
> Also, I just realized that Clone Hero only shows the the countdown and doesn't do the "Get Ready!" that RB4 does. Do you have a preference between the two styles?

I personally prefer what Clone Hero does, as it feels more polished and cohesive *because* it doesn't use the second textbox style. In YARG, the "Get Ready!" textbox would definitely feel a little out of place, as there's nothing else that uses that style of box except for the solo box.

CH also has the progress ring around the countdown circle, which is a nice little indicator and compliments the circle design.
